overlord/assertstate/validation_set_tracking.go

```diff
@@ -172,3 +181,79 @@ func EnforcedValidationSets(st *state.State) (*snapasserts.ValidationSets, error
 
 	return sets, err
 }
+
+// AddCurrentTrackingToValidationSetsHistory stores the current state of validation-sets
+// tracking on top of the validation sets history.
+func AddCurrentTrackingToValidationSetsHistory(st *state.State) error {
+	all, err := ValidationSets(st)
+	if err != nil {
+		return err
+	}
+	return addToValidationSetsHistory(st, all)
+}
+
+func addToValidationSetsHistory(st *state.State, currentTracking map[string]*ValidationSetTracking) error {
+	vshist, err := ValidationSetsHistory(st)
+	if err != nil {
+		return err
+	}
+	if len(currentTracking) == 0 && len(vshist) == 0 {
+		return nil
+	}
+
+	var matches bool
+	if len(vshist) > 0 {
+		// only add to the history if it's different than topmost entry
+		top := vshist[len(vshist)-1]
+		if len(top) == len(currentTracking) {
+			matches = true
+			for vskey, tr := range currentTracking {
+				prev, ok := top[vskey]
+				if !ok {
+					matches = false
+					break
+				}
+				if !prev.sameAs(tr) {
+					matches = false
+					break
+				}
+			}
+		}
+	}
+	if !matches {
+		vshist = append(vshist, currentTracking)
+	}
+	if len(vshist) > maxValidationSetsHistorySize {
+		vshist = vshist[len(vshist)-maxValidationSetsHistorySize:]
+	}
+	st.Set("validation-sets-history", &vshist)
+	return nil
+}
+
+// ValidationSetsHistoryTop returns the topmost validation sets tracking state from
+// the validations sets tracking history.
+func ValidationSetsHistoryTop(st *state.State) (map[string]*ValidationSetTracking, error) {
+	var vshist []*json.RawMessage
+	if err := st.Get("validation-sets-history", &vshist); err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	if len(vshist) == 0 {
+		return nil, nil
+	}
+	// decode just the topmost entry
```

Comment:
This is because the function that gets all entries uses passes []map[string]*ValidationSetTracking to st.Get, while the other one uses []*json.RawMessage in which case array entries remain in raw format; we use this trick in other places where we are only interested in specific records.

Agreed with unexporting ValidationSetsHistoryTop, it's used internally in the followup PR but doesn't have to be public I think.